go-controller/pkg/ovn/ovn.go

```diff
@@ -225,6 +267,66 @@ func (oc *DefaultNetworkController) removeLocalZonePod(pod *kapi.Pod, portInfo *
 	return nil
 }
 
+// removeRemoteZonePod tries to tear down a remote zone pod bits. It returns nil on success and error on failure;
+// failure indicates the pod tear down should be retried later.
+// It removes the remote pod ips from the namespace address set and if its an external gw pod, removes
+// its routes.
+func (oc *DefaultNetworkController) removeRemoteZonePod(pod *kapi.Pod) error {
+	annotation, err := util.UnmarshalPodAnnotation(pod.Annotations, types.DefaultNetworkName)
+	podDesc := fmt.Sprintf("pod %s/%s/%s", types.DefaultNetworkName, pod.Namespace, pod.Name)
+	if err != nil {
+		if util.IsAnnotationNotSetError(err) {
+			// if the annotation doesn’t exist, that’s not an error.
+			klog.V(5).Infof("No annotations on remote pod %s, no need to delete its ips from the namespace address set", podDesc)
+			return nil
+		}
+		return fmt.Errorf("unable to unmarshal pod annotations for %s: %w", podDesc, err)
+	}
+
+	podIfAddrs := annotation.IPs
+
+	// Remote the pod ips from the namespace address set.  Before that check if its a completed pod and
+	// make sure that the ips are not colliding with other pod
+	shouldRelease := true
+	if util.PodCompleted(pod) {
+		var needleIPs []net.IP
+		for _, podIPNet := range podIfAddrs {
+			needleIPs = append(needleIPs, podIPNet.IP)
+		}
+		collidingPod, err := oc.findPodWithIPAddresses(needleIPs)
+		if err != nil {
+			klog.Errorf("Unable to determine if completed remote pod IP is in use by another pod. "+
+				"Will not release pod %s/%s IP: %#v from namespace addressset. %w", pod.Namespace, pod.Name, podIfAddrs, err)
+			shouldRelease = false
+		}
+
+		if collidingPod != nil {
+			klog.Infof("Will not release IP address: %s for %s/%s from namespace address set. Detected another pod"+
+				" using this IP: %s/%s", util.JoinIPNetIPs(podIfAddrs, " "), pod.Namespace, pod.Name, collidingPod.Namespace, collidingPod.Name)
+			klog.Infof("Will not release IP address: %s for %s from remote pod's namespace address set. Detected another pod"+
+				" using this IP: %s/%s", util.JoinIPNetIPs(podIfAddrs, " "), podDesc, collidingPod.Namespace, collidingPod.Name)
+			shouldRelease = false
+		}
+	}
```

Comment:
It just looks to me that all this logic could be shared for local/remote pods?

Reply:
Ack